[RNMobile] Fix bug - multiple media selection adds only one image or video block

marecar3 opened the pull request:
## Description
This PR addresses an [issue](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1346)
where multiple media selection adds only one image or video block on Android devices and on iOS where the Force Touch app shortcut doesn't work properly selecting "New Photo Post".

Related [gutenberg-mobile PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/1383): 

## How has this been tested?

**[Android]**

Current behavior:

1. Add an image/video block
2. Add multiple images/videos from WordPress media library
3. Only 1 image/video block is added

Expected behavior:
1. Add an image/video block
2. Add multiple images/videos from WordPress media library
3. Multiple image/video blocks are added with the selected images/videos

**[iOS]**

This also seems to be the cause of an issue on WPiOS, where the Force Touch app shortcut doesn't work properly selecting "New Photo Post".

Steps:

1. Force Touch on the WPiOS app icon.
2. Select New Photo Post shortcut.
3. Make sure Gutenberg was opened as the editor.
4. Select an Image.
5. 💥 No image block was appended.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

mchowning committed:
[RNMobile] Insure tapping at end of post inserts at end

Previously, tapping at the end of the post would insert a block
immediately after the currently selected block. In addition, this commit
is cleaning out a few unusued props in the block-list file.

etoledom committed:
[RNMobile] Native mobile release v1.11.0 (#17181)

* [RNMobile] Fix crash when adding separator

* Build: remove global install of latest npm since we want to use the paired node/npm version (#17134)

* Build: remove global install of latest npm since we want to use the paired node/npm version
* Also update travis to remove --latest-npm flag

* [RNMobile] Try dark mode (iOS) (#17067)

* Adding dark mode component implemented on list and list block

* Adding DarkMode handling to RichText, ToolBar and SafeArea

* Mobile: Using DarkMode as HOC

* iOS DarkMode: Modified colors on block list and block container

* iOS DarkMode: Improved Header Toolbar colors

* iOS DarkMode: Removing background from buttons

* iOS DarkMode warning and unsupported

* iOS DarkMode: MediaPlaceholder

* iOS DarkMode: BottomSheets

* iOS DarkMode: Inserter

* iOS DarkMode: DefaultBlockAppender

* iOS DarkMode: PostTite

* Update hardcoded colors with variables

* iOS DarkMode: Fix bottom-sheet cell value color

* iOS DarkMode: More - PageBreak - Add Block Here

* iOS DarkMode: Better text color

* iOS Darkmode: Code block

* iOS DarkMode: HTML View

* iOS DarkMode: Improve colors on SafeArea

* Fix toolbar not avoiding keyboard regression

* Fix native unit tests

* Fix gutenberg-mobile unit tests

* Adding RNDarkMode mocks

* RNMobile: Fix crash when viewing HTML on iOS

* [RNMobile] Remove toolbar from html view

* [RNMobile] Fix MaxListenersExceededWarning caused by dark-mode event emitter (#17186)

* Fix MaxListenersExceededWarning caused by dark-mode event emitter

* Checking for setMaxListeners trying to avoid CI error

* Adding remove listener to DarkMode HOC

* DarkMode: Binding this.onModeChanged to `this`

* DarkMode: Adding conditional needed to pass UI Tests on CI

* Fix focus title on new posts regression (#17180)

* BottomSheet: Setting DashIcon color directly when theme is default (light) (#17193)

etoledom committed:
Activate Travis CI on rnmobile/master branch (#17229)

Tug committed:
Add native support for the MediaText block (#16305)

* First working version of the MediaText component for native mobile

* Fix adding a block to an innerblock list

* Disable mediaText on production

* MediaText native: improve editor visuals

* Move BlockToolbar from BlockList to Layout

* Remove BlockEditorProvider from BlockList and add native version of EditorProvider to Editor. Plus support InsertionPoint and BlockListAppender

* Update BlockMover for native to hide if locked or if it's the only block

* Make the vertical align button work, add more styling options for toolbar buttons

* Make sure registerCoreBlocks does not break in production

* Copy docblock comment from the web version for registerCoreBlocks

* Fix focusing on the media placeholder

* Only support adding image for now

* Update usage of MediaPlaceholder in MediaContainer

* Enable autoScroll for just the out most block list

* Fix JS Unit tests

* Roll back to IconButton refactor and fix tests

* Fix BlockVerticalAlignmentToolbar buttons style on mobile

* Fix thing for web and ensure ariaPressed is always passed down

* Use AriaPressed directly to style SVG on mobile

* Update snapshots

dratwas committed:
MediaUpload and MediaPlaceholder unify props (#17145)

lukewalczak committed:
Unify media placeholder and upload props within media-text (#17268)

geriux committed:
[RNMobile] Fix dismiss keyboard button for the post title (#17260)

etoledom committed:
Recover border colors (#17269)

lukewalczak committed:
Support group block on mobile (#17251)

* First working version of the MediaText component for native mobile

* Fix adding a block to an innerblock list

* Disable mediaText on production

* MediaText native: improve editor visuals

* Move BlockToolbar from BlockList to Layout

* Remove BlockEditorProvider from BlockList and add native version of EditorProvider to Editor. Plus support InsertionPoint and BlockListAppender

* Update BlockMover for native to hide if locked or if it's the only block

* Make the vertical align button work, add more styling options for toolbar buttons

* Make sure registerCoreBlocks does not break in production

* Copy docblock comment from the web version for registerCoreBlocks

* Fix focusing on the media placeholder

* Only support adding image for now

* Update usage of MediaPlaceholder in MediaContainer

* Enable autoScroll for just the out most block list

* Fix JS Unit tests

* Roll back to IconButton refactor and fix tests

* Fix BlockVerticalAlignmentToolbar buttons style on mobile

* Fix thing for web and ensure ariaPressed is always passed down

* Use AriaPressed directly to style SVG on mobile

* Update snapshots

* Support group block on mobile

* Extend shouldShowInsertionPoint condition to be false when group is selected

* Code refactor

* Update package-lock

lukewalczak committed:
Remove redundant bg color within button appender (#17325)

etoledom committed:
[RNMobile] DarkMode improvements (#17309)

* Remove the need to import `useStyle` and pass the theme prop on every instance that `withStyle` is used

* Implement dark-mode refactor on all components

* Fix broken native tests

* Fix default block appender background color on DarkMode

* DarkMode: Make `useStyle` a class function

daniloercoli committed:
[RNMobile] Add autosave to mobile apps (#17329)

* [RNMobile] Fix crash when adding separator

* Build: remove global install of latest npm since we want to use the paired node/npm version (#17134)

* Build: remove global install of latest npm since we want to use the paired node/npm version
* Also update travis to remove --latest-npm flag

* [RNMobile] Try dark mode (iOS) (#17067)

* Adding dark mode component implemented on list and list block

* Adding DarkMode handling to RichText, ToolBar and SafeArea

* Mobile: Using DarkMode as HOC

* iOS DarkMode: Modified colors on block list and block container

* iOS DarkMode: Improved Header Toolbar colors

* iOS DarkMode: Removing background from buttons

* iOS DarkMode warning and unsupported

* iOS DarkMode: MediaPlaceholder

* iOS DarkMode: BottomSheets

* iOS DarkMode: Inserter

* iOS DarkMode: DefaultBlockAppender

* iOS DarkMode: PostTite

* Update hardcoded colors with variables

* iOS DarkMode: Fix bottom-sheet cell value color

* iOS DarkMode: More - PageBreak - Add Block Here

* iOS DarkMode: Better text color

* iOS Darkmode: Code block

* iOS DarkMode: HTML View

* iOS DarkMode: Improve colors on SafeArea

* Fix toolbar not avoiding keyboard regression

* Fix native unit tests

* Fix gutenberg-mobile unit tests

* Adding RNDarkMode mocks

* RNMobile: Fix crash when viewing HTML on iOS

* [RNMobile] Remove toolbar from html view

* [RNMobile] Fix MaxListenersExceededWarning caused by dark-mode event emitter (#17186)

* Fix MaxListenersExceededWarning caused by dark-mode event emitter

* Checking for setMaxListeners trying to avoid CI error

* Adding remove listener to DarkMode HOC

* DarkMode: Binding this.onModeChanged to `this`

* DarkMode: Adding conditional needed to pass UI Tests on CI

* Fix focus title on new posts regression (#17180)

* BottomSheet: Setting DashIcon color directly when theme is default (light) (#17193)

* Add a preliminary version of the AutosaveMonitor for mobile that calls the "bridge" and asks the native side to save the content

* Add autosave mock function for tests

* Fix merge conflicts

* Fix lint

* Re-add autosave on mobile that was removed erroneously during import-merge from rnmobile/master

* Remove native variant of AutosaveMonitor and introduces changes at  editor store level

* Default to false for `isEditedPostAutosaveable` on mobile. There was a typo in the returing value on the previous commit.

* Make sure to consider edits to the Title when checking if auto-save is needed

* Fix lint

lukewalczak committed:
Add isAppender functionality on mobile (#17195)

* Add isAppender functionality on mobile

* refactor isAppender conditions

* Replace dropZoneUIOnly in favour of showMediaSelectionUI

* deprecate dropZoneUIOnly and add disableMediaSelection prop

* Update test

* Refactor tests and change prop name

* Remove redundant empty lines

* Refactor conditions inside MediaPlaceholder

* Update block-editor CHANGELOG

* Update packages/block-editor/CHANGELOG.md

Co-Authored-By: Grzegorz (Greg) Ziółkowski <grzegorz@gziolo.pl>

daniloercoli committed:
Autosave monitor - Make the mobile editor ping the native at each keystroke, since the deboucing logic is already well defined in the apps. (#17548)

Tug committed:
[RNMobile] Refactor Dark Mode HOC (#17552)

* [RNMobile] Refactor the Dark Mode HOC to fix naming antipatterns

* Fix lint errors

* Add .native.js suffix to usePreferredColorScheme

* Update usage of theme props renamed to preferredColorScheme

* Update usage of theme props renamed to preferredColorScheme

marecar3 committed:
Fix bug which results in presenting only one media block even multiple media files were selected

marecar3 committed:
Fixed lint errors

SergioEstevao committed:
Add missing heading levels to the UI (H4, H5, H6) (#17533)

lukewalczak committed:
Fix lint issue (#17598)

SergioEstevao committed:
Fix list filter on paste for RN mobile. (#17550)

* Fix method for RN mobile.

* Use array.From instead of slice.

* Remove comment and use Array.from directly

* Convert from NodeList spreadable to Array.from

* Fix lint errors.

* Fix documentation examples to use Array.from

* Add empty line.

geriux committed:
[RNMobile] Move MediaUploadPorgress to its own component folder (#17392)

* Move MediaUploadPorgress to its own component folder (native)

* MediaUploadProgress - Fix import to code standards

* MediaUploadProgress readme

* Mobile - MediaUploadProgress README update

marecar3 committed:
Merge branch 'rnmobile/master' into rnmobile/1346_Multiple-media-selection-adds-only-one-image-or-video-block

marecar3 committed:
Merge branch 'master' into rnmobile/1346_Multiple-media-selection-adds-only-one-image-or-video-block

# Conflicts:
#	packages/block-editor/src/components/media-placeholder/README.md
#	packages/block-editor/src/components/media-upload/README.md
#	packages/compose/src/hooks/use-preferred-color-scheme/index.native.js
#	packages/editor/src/components/provider/index.native.js
#	test/native/setup.js

marecar3 committed:
removed merge tags

marecar3 committed:
removed doubled lines

etoledom commented on packages/editor/src/components/provider/index.native.js:
```diff
@@ -63,6 +64,16 @@ class NativeEditorProvider extends Component {
+				[ payload.mediaType === 'image' ? 'url' : 'src' ]: payload.mediaUrl,
+			} );
+
+			this.props.insertBlock( newBlock, this.props.blockCount );
```
The only problem with this is that all images but the first one will be added at the end of the post.

Not sure if at this point we have enough information to add each new image when it corresponds? 🤔 
For sure can be an improvement for the future.

etoledom reviewed (commented):
Thank you @marecar3 for taking care of this!

Tested on WPAndroid and WPiOS and this solves the issues in both platforms 🎉 

⚠️ There are some extra commits showing, probably it would be good to fix the history before merging.

**WPiOS:**
On iOS for some reason the resulting image is a very low quality one. This for sure is an issue on the Native side not related to this PR, but still wanted to highlight it.

![ios-shortcut](https://user-images.githubusercontent.com/9772967/65867357-89db2f80-e376-11e9-8d9f-8286cb4ee35f.png)

**WPAndroid:**
On Android, there's a small inconvenience that, if I want to add the I mages at the beginning of the post (or anywhere that is not the end), the first image block is positioned correctly, but all the rest are added at the end of the block.

Not sure how important is this problem, but I'd me OK to make a ticket out of this to be handled in the future (cc @hypest @pinarol ):

![add-multiple](https://user-images.githubusercontent.com/9772967/65867485-ce66cb00-e376-11e9-94da-efd7884421cd.png)

marecar3 committed:
Merge branch 'master' into rnmobile/1346_Multiple-media-selection-adds-only-one-image-or-video-block

marecar3 committed:
Ensure to show multiple images block bellow selected image block

marecar3 commented:
Hey @etoledom, thanks for the good review!

I fixed an issue when images are shown at the end of the post, not they will be shown bellow selected block which should be image/video block. Let me know what do you think about that approach. Thanks.

Regarding the low-quality image, I have experienced similar on Android but just managed to reproduce it once.

PR is ready for another iteration.

etoledom commented on packages/editor/src/components/provider/index.native.js:
```diff
@@ -151,12 +166,21 @@ export default compose( [
 			blocks: getEditorBlocks(),
 			title: getEditedPostAttribute( 'title' ),
 			getEditedPostContent,
+			selectedBlockIndex: selectedBlockIndex,
```
Here you can avoid repeating `selectedBlockIndex`

marecar3 replied:
Good catch! Fixed!

etoledom reviewed (commented):
Thank you for the update @marecar3 !

I just caught an issue with this new strategy, while it does solve the problem previously described, now the images are appended to the post on the reversed order from the selection.

My guess is that since we are appending always at the selected block index, every next image is added always at the top index.

![multiple_media](https://user-images.githubusercontent.com/9772967/65962658-7359d480-e459-11e9-927e-ebbe0bdfb46a.png)

Again, I'm not sure if there's an easy way to make this work properly without digging into the code more deeply, and I don't recall what was the behavior previously. 🤔

marecar3 committed:
Just pass selectedBlockIndex constant as a prop instead of assignment

marecar3 committed:
Fix wrong order of inserted image blocks

marecar3 commented:
Hey, @etoledom thanks for the good review!

I saw the old code and took the solution from here: https://github.com/wordpress-mobile/gutenberg-mobile/blob/cb93dd24e50855f15f43f2ce00b2eaa76c1bd8ef/src/block-management/block-manager.js#L115

I tested and didn't find any problems. Let me know your opinion on this one.

PR is ready for another iteration of testing. Thanks!!!

etoledom reviewed (approved):
Tested on both iOS and Android again.
iOS keep working as mentioned before (fixed the original issue 🎉 )
Android is not working as expected! Nice job @marecar3 🥇

etoledom commented:
There are a couple of lint errors:
```
[0] /home/travis/build/WordPress/gutenberg/packages/editor/src/components/provider/index.native.js
835[0]   179:1   error  Trailing spaces not allowed  no-trailing-spaces
836[0]   187:45  error  Missing trailing comma       comma-dangle
```

The autosave tests error (as far as I understand) are safe to ignore for now.

marecar3 committed:
Fixed lint errors

marecar3 committed:
Merge branch 'master' into rnmobile/1346_Multiple-media-selection-adds-only-one-image-or-video-block

marecar3 commented:
Hey @etoledom, thanks for the approve and for good review 👍 

I have fixed the lint tests and updated release notes. 
When tests pass I will merge this PR.
